launcher/archive/ArchiveWriter.cpp

```diff
@@ -0,0 +1,211 @@
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+#include "ArchiveWriter.h"
+#include <archive.h>
+#include <archive_entry.h>
+#include <sys/stat.h>
+
+#include <QFile>
+#include <QFileInfo>
+
+#include <memory>
+
+namespace MMCZip {
+
+ArchiveWriter::ArchiveWriter(const QString& archiveName) : m_filename(archiveName) {}
+
+ArchiveWriter::~ArchiveWriter()
+{
+    close();
+}
+
+bool ArchiveWriter::open()
+{
+    if (m_filename.isEmpty()) {
+        qCritical() << "Archive m_filename not set.";
+        return false;
+    }
+
+    m_archive = archive_write_new();
+    if (!m_archive) {
+        qCritical() << "Archive not initialized.";
+        return false;
+    }
+
+    archive_write_set_format_zip(m_archive);
+
+    if (archive_write_set_options(m_archive, "hdrcharset=UTF-8") != ARCHIVE_OK) {
+        qCritical() << "Failed to open archive file:" << m_filename << "-" << archive_error_string(m_archive);
+        return false;
+    }
+
+    auto archiveNameUtf8 = m_filename.toUtf8();
+    if (archive_write_open_filename(m_archive, archiveNameUtf8.constData()) != ARCHIVE_OK) {
+        qCritical() << "Failed to open archive file:" << m_filename << "-" << archive_error_string(m_archive);
+        return false;
+    }
+
+    return true;
+}
+
+bool ArchiveWriter::close()
+{
+    bool success = true;
+    if (m_archive) {
+        if (archive_write_close(m_archive) != ARCHIVE_OK) {
+            qCritical() << "Failed to close archive" << m_filename << "-" << archive_error_string(m_archive);
+            success = false;
+        }
+        if (archive_write_free(m_archive) != ARCHIVE_OK) {
+            qCritical() << "Failed to free archive" << m_filename << "-" << archive_error_string(m_archive);
+            success = false;
+        }
+        m_archive = nullptr;
+    }
+    return success;
+}
+
+bool ArchiveWriter::addFile(const QString& fileName, const QString& fileDest)
+{
+    QFileInfo fileInfo(fileName);
+    if (!fileInfo.exists()) {
+        qCritical() << "File does not exists:" << fileInfo.filePath();
+        return false;
+    }
+
+    std::unique_ptr<archive_entry, void (*)(archive_entry*)> entry_ptr(archive_entry_new(), archive_entry_free);
+    auto entry = entry_ptr.get();
+    if (!entry) {
+        qCritical() << "Failed to create archive entry";
+        return false;
+    }
+
+    auto fileDestUtf8 = fileDest.toUtf8();
+    archive_entry_set_pathname(entry, fileDestUtf8.constData());
+
+    QByteArray utf8 = fileInfo.absoluteFilePath().toUtf8();
+    const char* cpath = utf8.constData();
+    struct stat st;
+    if (stat(cpath, &st) == 0) {
+        archive_entry_copy_stat(entry, &st);
+    }
+    archive_entry_set_perm(entry, fileInfo.permissions() & 0777);
```

Comment:
```suggestion
    archive_entry_set_perm(entry, fileInfo.permissions());
```
Pretty sure masking this value would lose permission attributes for NTFS/Windows stuff

Reply:
nope this one only loses SUID/SGID/sticky bits not the rwx permissions

Reply:
but I see no reason not to follow your suggestion